### PR TITLE
filter: AWS Lambda filter (initial version)

### PR DESCRIPTION
**Pull request by marcomagdy**
This filter transform HTTP requests to AWS Lambda invocations.

The filter supports pass-through only. Meaning, the request body
is passed to Lambda as is. Note: Lambda requires the request to be in
JSON format.

In a later iteration, we'll wrap the headers the body in a JSON string
before passing it to Lambda.

The filter requires the ARN of the Lambda function and supports
per-filter-config. When the per-filter configuration is used, the target
cluster must be tagged with specific metadata. This indicates to the
filter whether to process the request or to skip it.

Lambda supports two invocation modes:
- Synchronous (Request-Response)
- Asynchronous (Event)

This initial version of the filter supports the synchronous mode only.
In a later iteration I'll add support for the asynchronous (Event-based)
version.

Signed-off-by: Marco Magdy <mmagdy@gmail.com>

### CODE REVIEW

File: source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+  }
+
+  if (end_stream) {
+    setHeaders(headers, arn_->functionName());
+    // We must NOT sign this header, because it causes signature miscalculation at Lambda's side.
+    // We _could_ remove the header, sign the request, and then tack it back on. But, I don't see
+    // value in that header when targeting AWS Lambda.
+    if (headers.ForwardedProto()) {
+      headers.remove(Http::Headers::get().ForwardedProto);
+    }
```
Comment (marcomagdy) [inline source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc]: This can be removed once we update the aws sigv4 signer to do that.

File: api/envoy/config/filter/http/aws_lambda/v2alpha/aws_lambda.proto [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.config.filter.http.aws_lambda.v2alpha";
+option java_outer_classname = "AwsLambdaProto";
+option java_multiple_files = true;
+option (udpa.annotations.file_migrate).move_to_package =
+    "envoy.extensions.filters.http.aws_lambda.v3";
+option (udpa.annotations.file_status).work_in_progress = true;
+
+// [#protodoc-title: AwsLambda]
```
Comment (mattklein123) [inline api/envoy/config/filter/http/aws_lambda/v2alpha/aws_lambda.proto]: nit: AWS Lambda

File: docs/root/configuration/http/http_filters/aws_lambda_filter.rst [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+.. attention::
+
+  The AWS Lambda filter is currently under active development.
+
+The HTTP AWS Lambda filter is used to trigger an AWS Lambda function from a standard HTTP/1.x or HTTP/2 request.
+It supports a few options to control whether to pass through the HTTP request payload as is or to wrap it in a JSON
+schema.
+
+If ``payload_passthrough`` is set to ``true``, then the payload is sent to Lambda without any transformations.
```
Comment (mattklein123) [inline docs/root/configuration/http/http_filters/aws_lambda_filter.rst]: Please ref link all config options (or at least the first one)

File: source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    return Http::FilterDataStatus::Continue;
+  }
+
+  if (end_stream) {
+    setHeaders(*headers_, arn_->functionName());
+    auto& hashing_util = Envoy::Common::Crypto::UtilitySingleton::get();
+    const Buffer::Instance& decoding_buffer = *decoder_callbacks_->decodingBuffer();
+    const auto hash = Hex::encode(hashing_util.getSha256Digest(decoding_buffer));
+    if (headers_->ForwardedProto()) {
+      headers_->remove(Http::Headers::get().ForwardedProto); // We must NOT sign this header.
```
Comment (mattklein123) [inline source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc]: Just use the O(1) remove function here.

File: source/extensions/filters/http/aws_lambda/aws_lambda_filter.h [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  const std::string& resourceType() const { return resource_type_; }
+  const std::string& functionName() const { return function_name_; }
+
+private:
+  std::string partition_;
+  std::string service_;
+  std::string region_;
+  std::string account_id_;
+  std::string resource_type_;
+  std::string function_name_; // resource_id
```
Comment (mattklein123) [inline source/extensions/filters/http/aws_lambda/aws_lambda_filter.h]: nit: const. Same below/elsewhere where possible.

File: source/extensions/filters/http/aws_lambda/config.cc [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  }
+  throw EnvoyException(fmt::format("Invalid ARN: {}", arn));
+}
+} // namespace
+
+Http::FilterFactoryCb AwsLambdaFilterFactory::createFilterFactoryFromProtoTyped(
+    const envoy::extensions::filters::http::aws_lambda::v3::Config& proto_config,
+    const std::string& stats_prefix, Server::Configuration::FactoryContext& context) {
+
+  (void)stats_prefix;
```
Comment (mattklein123) [inline source/extensions/filters/http/aws_lambda/config.cc]: nit: just delete var name above or use UNREFERNCED_PARAMETER

File: test/extensions/common/aws/mocks.h [hunk 7]
```diff
@@ -12,19 +12,14 @@ namespace Aws {
 
 class MockCredentialsProvider : public CredentialsProvider {
 public:
-  MockCredentialsProvider();
-  ~MockCredentialsProvider() override;
```
Comment (mattklein123) [inline test/extensions/common/aws/mocks.h]: Please revert this change. This is a perf optimization recommended by gmock.

File: docs/root/configuration/http/http_filters/aws_lambda_filter.rst [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+*Note*: This means you lose access to all the HTTP headers in the Lambda function.
+
+However, if ``payload_passthrough`` is set to ``false``, then the HTTP request is transformed to a JSON payload with the following
```
Comment (mattklein123) [inline docs/root/configuration/http/http_filters/aws_lambda_filter.rst]: This is not implemented yet, right? I'm a little confused about what is implemented right now versus what is not. Should you just remove the docs from this PR until it's actually implemented?

Comment (marcomagdy) [reply docs/root/configuration/http/http_filters/aws_lambda_filter.rst]: I added it to get a review on the examples mainly. I can remove the parts of the documentation (and examples) that speak to features that are not implemented yet.

If you prefer to review documentation once at the end, that works too. LMK.

File: source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc (hunk 4 above)
Comment (marcomagdy) [reply source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc]: sure

File: source/extensions/filters/http/aws_lambda/aws_lambda_filter.h (hunk 5 above)
Comment (marcomagdy) [reply source/extensions/filters/http/aws_lambda/aws_lambda_filter.h]: I can't mark those members `const` because I need to update `arn_` member. FWIW, this class started as a simple POD with `const` members.

File: source/extensions/filters/http/aws_lambda/config.cc (hunk 6 above)
Comment (marcomagdy) [reply source/extensions/filters/http/aws_lambda/config.cc]: I'll remove it.

File: test/extensions/common/aws/mocks.h (hunk 7 above)
Comment (marcomagdy) [reply test/extensions/common/aws/mocks.h]: I'll revert it. But I couldn't find anything in the gtest/gmock documentation that speak to that guideline. 
Please point me to it if you can. I'm curious.

File: docs/root/configuration/http/http_filters/aws_lambda_filter.rst (hunk 3 above)
Comment (marcomagdy) [reply docs/root/configuration/http/http_filters/aws_lambda_filter.rst]: Sure.

File: test/extensions/common/aws/mocks.h (hunk 7 above)
Comment (mattklein123) [reply test/extensions/common/aws/mocks.h]: https://github.com/google/googletest/blob/master/googlemock/docs/cook_book.md#making-the-compilation-faster

**Comment by marcomagdy** [comment]
cc @mattklein123
